proposals/destructuring-in-parameters.md

```diff
@@ -0,0 +1,156 @@
+
+NOTE: Support in parameters of functions (declarations and expressions), constructors and setters will be discussed for future-proof design, but is not proposed to be implemented at this time.
+
+## Use cases
+
+To be supported now:
+
+``` kotlin
+// decomposing pairs in a lambda
+listOfPairs.map {
+  (a, b) -> a + b
+}
+```
+
+May be supported later:
+
+``` kotlin
+// decompose a parameter:
+fun foo((a, b): Pair<Int, String>, c: Bar)
+// can be called as
+foo(pair, bar)
+
+// decompose a constructor parameter
+class C(val (a, b): Pair<Int, String>) {}
+```
+
+## Syntax
+
+We allow parentheses in lambda parameter lists:
+
+In Kotlin 1.0 the syntax for a lambda is:
+
+```
+functionLiteral
+  : "{" statements "}"
+  : "{" (SimpleName (":" type)?){","} "->" statements "}"
+  ;
+```
+
+Here we propose to extend it:
+
+```
+functionLiteral
+  : "{" statements "}"
+  : "{" lambdaParameter{","} "->" statements "}"
+  ;
+  
+lambdaParameter
+  : variableDeclarationEntry
+  : multipleVariableDeclarations (":" type)? 
+  ;
+  
+// already defined, shown for reference
+  
+variableDeclarationEntry
+  : SimpleName (":" type)?
+  ;
+
+multipleVariableDeclarations
+  : "(" variableDeclarationEntry{","} ")"
+  ;  
+```
+
+> Note: while modifiers in front of lambda parameters are mentioned in the grammar for 1.0, they are not actually supported by the parser.
+> If we ever support modifiers here, we should allow them in front of individual components, but for now, the actual rule should be without them  
+
+Examples:
+
+``` kotlin
+{ a -> ... } // one parameter
+{ a, b -> ... } // two parameters
+{ (a, b) -> ... } // a destructured pair
+{ (a, b), c -> ... } // a destructured pair and another parameter
+```
+
+> Question: Should we support nested destructuring for parameters?
+> ```
+{ ((a, b), c) -> ... } // a destructured pair whose first component is a pair
+```
+> NOTE: this is not supported for destructuring val's and var's 
+
+A type may be specified for the whole destructured parameter, for its individual components (independently), or both: 
+
+``` kotlin
+{ (a, b: B) -> ... }
+{ (a, b): Pair<A, B> -> ... }
+{ (a: A, b): Pair<A, B> -> ... }
+```
+
+> Note: no changes to the syntax of function types is needed since the syntax for call sites (where a lambda is invoked) is not changed.
+  
+### Semantics and checks
+
+Semantically, destructured parameters work as if they are normal parameters and a destructuring assignments are performed before the body of the lambda, in the left-to-right order:
+  
+``` kotlin
+{ 
+  (a, b), (c, d) -> 
+  body() 
+}
+```
+
+is translated to
+
+```
+{ 
+  _p1, _p2 ->
+  val (a, b) = _p1
+  val (c, d) = _p2 
+  body() 
+}
+```
+
+**Typing rule**: Types of the components must be compatible with component-functions (`componentN()`) resolvable on corresponding parameters. When such functions can not be applied or their return types are not assignable to declared types of components, an error is reported.
```

Comment:
I suppose that it should work. In general, we should do the following: compute type of `it`, then resolve statements: `val a(: Type) = it.component1()`.